ratpack-core/src/main/java/ratpack/exec/util/Batch.java

```diff
@@ -0,0 +1,365 @@
+import ratpack.api.Nullable;
+import ratpack.exec.ExecResult;
+import ratpack.exec.Execution;
+import ratpack.exec.Promise;
+import ratpack.exec.Result;
+import ratpack.func.Action;
+import ratpack.func.BiAction;
+import ratpack.stream.Streams;
+import ratpack.stream.TransformablePublisher;
+import ratpack.stream.internal.BufferingPublisher;
+import ratpack.util.Types;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A mechanism for processing batches of jobs as {@link Promise promises}.
+ * <p>
+ * A batch is created via the {@link #of(Iterable)} method or one of its overloads.
+ * This type is effectively a builder for how to process the jobs.
+ * Some methods configure how the processing is to be done, and others initiate the processing.
+ * <p>
+ * Batches are immutable, reusable and are used in a fluent manner.
+ *
+ * <pre class="java">{@code
+ * import ratpack.exec.Promise;
+ * import ratpack.exec.util.Batch;
+ * import ratpack.test.exec.ExecHarness;
+ *
+ * import java.util.ArrayList;
+ * import java.util.List;
+ *
+ * import static org.junit.Assert.assertEquals;
+ *
+ * public class Example {
+ *   public static void main(String[] args) throws Exception {
+ *     int sum = ExecHarness.yieldSingle(e -> {
+ *
+ *       List<Promise<Integer>> jobs = new ArrayList<>();
+ *       for (int i = 0; i < 100; i++) {
+ *         jobs.add(Promise.value(i));
+ *       }
+ *
+ *       return Batch.of(jobs)
+ *         .parallel()
+ *         .publisher()
+ *         .reduce(0, (r, i) -> r + i);
+ *
+ *     }).getValueOrThrow();
+ *
+ *     assertEquals(sum, 4950);
+ *   }
+ * }
+ * }</pre>
+ *
+ * @param <T> the type of item in the batch
+ * @since 1.4
+ * @see #yieldAll()
+ * @see #publisher()
+ */
+public final class Batch<T> {
+
+  private final Iterable<? extends Promise<T>> jobs;
+  private final Action<? super Execution> execInit;
+
+  private Batch(Iterable<? extends Promise<T>> jobs, @Nullable Action<? super Execution> execInit) {
+    this.jobs = jobs;
+    this.execInit = execInit;
+  }
+
+  /**
+   * Creates a new batch of the given promises.
+   * <p>
+   * The given iterable will not be consumed until processing begins.
+   *
+   * @param jobs the batch jobs
+   * @param <T> the type of item produced by each job
+   * @return a new batch
+   */
+  public static <T> Batch<T> of(Iterable<? extends Promise<T>> jobs) {
+    return new Batch<>(jobs, null);
+  }
+
+  /**
+   * Creates a new batch of the given promises.
+   *
+   * @param jobs the batch jobs
+   * @param <T> the type of item produced by each job
+   * @return a new batch
+   */
+  @SafeVarargs
+  @SuppressWarnings("varargs")
+  public static <T> Batch<T> of(Promise<T>... jobs) {
+    return new Batch<>(Arrays.asList(jobs), null);
+  }
+
```

Comment:
You could also add static methods Batch.parallel(jobs) and Batch.parallel(jobs, execInit) as a convenience method to Batch.of(jobs).parallel() and Batch.of(jobs).parallel(execInit). This might make the parallelization features more visible.